library/std/src/sys/unix/args.rs

```diff
@@ -127,21 +116,16 @@ mod imp {
         init_wrapper
     };
 
-    pub unsafe fn cleanup() {
-        let _guard = LOCK.lock();
-        ARGC.store(0, Ordering::Relaxed);
-        ARGV.store(ptr::null_mut(), Ordering::Relaxed);
-    }
-
     pub fn args() -> Args {
         Args { iter: clone().into_iter() }
     }
 
     fn clone() -> Vec<OsString> {
         unsafe {
-            let _guard = LOCK.lock();
-            let argc = ARGC.load(Ordering::Relaxed);
+            // Load ARGC and ARGV without a lock. If the store to either ARGV or
+            // ARGC isn't visible yet, we'll return an empty argument list.
             let argv = ARGV.load(Ordering::Relaxed);
+            let argc = if argv.is_null() { 0 } else { ARGC.load(Ordering::Relaxed) };
```

Comment:
Don't you need to do an acquire load (and a release store elsewhere) to ensure that the non-atomic load later (`*argv.offset(i)`) does not cause a data race?

Reply:
I assume by this point the memory pointed to by argv is already ready to be loaded from without synchronization, since that's how C code would use argv.